File: bindings/php/tests/MullamaTest.php
<?php

declare(strict_types=1);

namespace Mullama\Tests;

use Mullama\Mullama;
use Mullama\Model;
use Mullama\Context;
use Mullama\SamplerParams;
use Mullama\EmbeddingGenerator;
use PHPUnit\Framework\TestCase;

/**
 * Tests for the Mullama PHP bindings.
 *
 * Note: Some tests require a model file. Set MULLAMA_TEST_MODEL environment variable.
 */
final class MullamaTest extends TestCase
{
    private static ?string $testModelPath = null;
    private static bool $modelAvailable = false;

    public static function setUpBeforeClass(): void
    {
        self::$testModelPath = getenv('MULLAMA_TEST_MODEL') ?: null;
        self::$modelAvailable = self::$testModelPath && file_exists(self::$testModelPath);
    }

    // === Version and System Info Tests ===

    public function testVersion(): void
    {
        $version = Mullama::version();
        $this->assertIsString($version);
        $this->assertEquals('0.1.0', $version);
    }

    public function testSystemInfo(): void
    {
        $this->markTestSkipped('FFI initialization required');
        // $info = Mullama::systemInfo();
        // $this->assertIsString($info);
        // $this->assertNotEmpty($info);
    }

    public function testSupportsGpuOffload(): void
    {
        $this->markTestSkipped('FFI initialization required');
        // $result = Mullama::supportsGpuOffload();
        // $this->assertIsBool($result);
    }

    public function testMaxDevices(): void
    {
        $this->markTestSkipped('FFI initialization required');
        // $devices = Mullama::maxDevices();
        // $this->assertIsInt($devices);
        // $this->assertGreaterThanOrEqual(1, $devices);
    }

    // === Sampler Params Tests ===

    public function testDefaultSamplerParams(): void
    {
        $params = new SamplerParams();
        $this->assertEquals(0.8, $params->temperature, '', 0.01);
        $this->assertEquals(40, $params->topK);
        $this->assertEquals(0.95, $params->topP, '', 0.01);
    }

    public function testCustomSamplerParams(): void
    {
        $params = new SamplerParams([
            'temperature' => 0.5,
            'topK' => 20,
            'topP' => 0.8,
        ]);
        $this->assertEquals(0.5, $params->temperature, '', 0.01);
        $this->assertEquals(20, $params->topK);
        $this->assertEquals(0.8, $params->topP, '', 0.01);
    }

    public function testGreedySamplerParams(): void
    {
        $params = SamplerParams::greedy();
        $this->assertEquals(0.0, $params->temperature, '', 0.01);
        $this->assertEquals(1, $params->topK);
    }

    public function testCreativeSamplerParams(): void
    {
        $params = SamplerParams::creative();
        $this->assertGreaterThan(1.0, $params->temperature);
        $this->assertGreaterThan(40, $params->topK);
    }

    public function testPreciseSamplerParams(): void
    {
        $params = SamplerParams::precise();
        $this->assertLessThan(0.5, $params->temperature);
        $this->assertLessThan(40, $params->topK);
    }

    // === Cosine Similarity Tests ===

    public function testCosineSimilarityIdentical(): void
    {
        $a = [1.0, 0.0, 0.0];
        $b = [1.0, 0.0, 0.0];
        $sim = EmbeddingGenerator::cosineSimilarity($a, $b);
        $this->assertEquals(1.0, $sim, '', 0.001);
    }

    public function testCosineSimilarityOrthogonal(): void
    {
        $a = [1.0, 0.0, 0.0];
        $b = [0.0, 1.0, 0.0];
        $sim = EmbeddingGenerator::cosineSimilarity($a, $b);
        $this->assertEquals(0.0, $sim, '', 0.001);
    }

    public function testCosineSimilarityOpposite(): void
    {
        $a = [1.0, 0.0, 0.0];
        $b = [-1.0, 0.0, 0.0];
        $sim = EmbeddingGenerator::cosineSimilarity($a, $b);
        $this->assertEquals(-1.0, $sim, '', 0.001);
    }

    public function testCosineSimilarityRandom(): void
    {
        $a = array_map(fn() => mt_rand(-100, 100) / 100, range(0, 127));
        $b = array_map(fn() => mt_rand(-100, 100) / 100, range(0, 127));
        $sim = EmbeddingGenerator::cosineSimilarity($a, $b);
        $this->assertGreaterThanOrEqual(-1.0, $sim);
        $this->assertLessThanOrEqual(1.0, $sim);
    }

    public function testCosineSimilarityMismatchedLength(): void
    {
        $this->expectException(\RuntimeException::class);
        $a = [1.0, 0.0];
        $b = [1.0, 0.0, 0.0];
        EmbeddingGenerator::cosineSimilarity($a, $b);
    }

    // === Model Loading Tests (require FFI and model) ===

    public function testLoadNonexistentModel(): void
    {
        $this->markTestSkipped('FFI initialization required');
        // $this->expectException(\RuntimeException::class);
        // Model::load('/nonexistent/path/model.gguf');
    }

    public function testLoadModel(): void
    {
        if (!self::$modelAvailable) {
            $this->markTestSkipped('No test model available');
        }

        $model = Model::load(self::$testModelPath);
        $this->assertGreaterThan(0, $model->nVocab());
        $this->assertGreaterThan(0, $model->nEmbd());
        $model->free();
    }

    public function testModelProperties(): void
    {
        if (!self::$modelAvailable) {
            $this->markTestSkipped('No test model available');
        }

        $model = Model::load(self::$testModelPath);

        $this->assertGreaterThan(0, $model->nCtxTrain());
        $this->assertGreaterThan(0, $model->nEmbd());
        $this->assertGreaterThan(0, $model->nVocab());
        $this->assertGreaterThan(0, $model->nLayer());
        $this->assertGreaterThan(0, $model->nHead());
        $this->assertGreaterThan(0, $model->size());
        $this->assertGreaterThan(0, $model->nParams());
        $this->assertGreaterThanOrEqual(0, $model->tokenBos());
        $this->assertGreaterThanOrEqual(0, $model->tokenEos());

        $model->free();
    }

    public function testTokenization(): void
    {
        if (!self::$modelAvailable) {
            $this->markTestSkipped('No test model available');
        }

        $model = Model::load(self::$testModelPath);

        $text = 'Hello, world!';
        $tokens = $model->tokenize($text);

        $this->assertIsArray($tokens);
        $this->assertNotEmpty($tokens);

        $decoded = $model->detokenize($tokens);
        $this->assertIsString($decoded);

        $model->free();
    }

    // === Context Tests (require FFI and model) ===

    public function testCreateContext(): void
    {
        if (!self::$modelAvailable) {
            $this->markTestSkipped('No test model available');
        }

        $model = Model::load(self::$testModelPath);
        $ctx = new Context($model);

        $this->assertGreaterThan(0, $ctx->nCtx());
        $this->assertGreaterThan(0, $ctx->nBatch());

        $ctx->free();
        $model->free();
    }

    public function testContextWithParams(): void
    {
        if (!self::$modelAvailable) {
            $this->markTestSkipped('No test model available');
        }

        $model = Model::load(self::$testModelPath);
        $ctx = new Context($model, ['nCtx' => 512, 'nBatch' => 256]);

        $this->assertEquals(512, $ctx->nCtx());
        $this->assertEquals(256, $ctx->nBatch());

        $ctx->free();
        $model->free();
    }

    public function testGenerateText(): void
    {
        if (!self::$modelAvailable) {
            $this->markTestSkipped('No test model available');
        }

        $model = Model::load(self::$testModelPath);
        $ctx = new Context($model, ['nCtx' => 256]);

        $params = SamplerParams::greedy();
        $text = $ctx->generate('Hello', 10, $params);

        $this->assertIsString($text);
        $this->assertNotEmpty($text);

        $ctx->free();
        $model->free();
    }

    public function testClearCache(): void
    {
        if (!self::$modelAvailable) {
            $this->markTestSkipped('No test model available');
        }

        $model = Model::load(self::$testModelPath);
        $ctx = new Context($model);

        // Should not throw
        $ctx->clearCache();

        $ctx->free();
        $model->free();
    }

    // === Embedding Generator Tests (require FFI and model) ===

    public function testCreateEmbeddingGenerator(): void
    {
        if (!self::$modelAvailable) {
            $this->markTestSkipped('No test model available');
        }

        $model = Model::load(self::$testModelPath);
        $gen = new EmbeddingGenerator($model);

        $this->assertGreaterThan(0, $gen->nEmbd());

        $gen->free();
        $model->free();
    }

    public function testEmbedText(): void
    {
        if (!self::$modelAvailable) {
            $this->markTestSkipped('No test model available');
        }

        $model = Model::load(self::$testModelPath);
        $gen = new EmbeddingGenerator($model);

        $embedding = $gen->embed('Hello, world!');

        $this->assertIsArray($embedding);
        $this->assertCount($gen->nEmbd(), $embedding);

        $gen->free();
        $model->free();
    }

    public function testEmbedBatch(): void
    {
        if (!self::$modelAvailable) {
            $this->markTestSkipped('No test model available');
        }

        $model = Model::load(self::$testModelPath);
        $gen = new EmbeddingGenerator($model);

        $texts = ['Hello', 'World', 'Test'];
        $embeddings = $gen->embedBatch($texts);

        $this->assertCount(3, $embeddings);
        foreach ($embeddings as $emb) {
            $this->assertIsArray($emb);
            $this->assertCount($gen->nEmbd(), $emb);
        }

        $gen->free();
        $model->free();
    }
}

File: bindings/go/mullama_test.go
package mullama

import (
	"math"
	"os"
	"testing"
)

var testModelPath = os.Getenv("MULLAMA_TEST_MODEL")
var modelAvailable = testModelPath != "" && fileExists(testModelPath)

func fileExists(path string) bool {
	_, err := os.Stat(path)
	return err == nil
}

func TestVersion(t *testing.T) {
	v := Version()
	if v == "" {
		t.Error("Version should not be empty")
	}
	if v != "0.1.0" {
		t.Errorf("Expected version 0.1.0, got %s", v)
	}
}

func TestSystemInfo(t *testing.T) {
	info := SystemInfo()
	if info == "" {
		t.Error("SystemInfo should not be empty")
	}
}

func TestSupportsGPUOffload(t *testing.T) {
	// Just ensure it doesn't panic
	_ = SupportsGPUOffload()
}

func TestMaxDevices(t *testing.T) {
	devices := MaxDevices()
	if devices < 1 {
		t.Errorf("MaxDevices should be at least 1, got %d", devices)
	}
}

func TestDefaultModelParams(t *testing.T) {
	params := DefaultModelParams()
	if params.NGPULayers != 0 {
		t.Errorf("Default NGPULayers should be 0, got %d", params.NGPULayers)
	}
	if !params.UseMmap {
		t.Error("Default UseMmap should be true")
	}
	if params.UseMlock {
		t.Error("Default UseMlock should be false")
	}
	if params.VocabOnly {
		t.Error("Default VocabOnly should be false")
	}
}

func TestDefaultContextParams(t *testing.T) {
	params := DefaultContextParams()
	if params.NCtx != 0 {
		t.Errorf("Default NCtx should be 0, got %d", params.NCtx)
	}
	if params.NBatch != 2048 {
		t.Errorf("Default NBatch should be 2048, got %d", params.NBatch)
	}
	if params.NThreads <= 0 {
		t.Error("Default NThreads should be positive")
	}
	if params.Embeddings {
		t.Error("Default Embeddings should be false")
	}
}

func TestDefaultSamplerParams(t *testing.T) {
	params := DefaultSamplerParams()
	if math.Abs(float64(params.Temperature-0.8)) > 0.01 {
		t.Errorf("Default Temperature should be 0.8, got %f", params.Temperature)
	}
	if params.TopK != 40 {
		t.Errorf("Default TopK should be 40, got %d", params.TopK)
	}
	if math.Abs(float64(params.TopP-0.95)) > 0.01 {
		t.Errorf("Default TopP should be 0.95, got %f", params.TopP)
	}
}

func TestGreedySamplerParams(t *testing.T) {
	params := GreedySamplerParams()
	if params.Temperature != 0.0 {
		t.Errorf("Greedy Temperature should be 0.0, got %f", params.Temperature)
	}
	if params.TopK != 1 {
		t.Errorf("Greedy TopK should be 1, got %d", params.TopK)
	}
}

func TestCreativeSamplerParams(t *testing.T) {
	params := CreativeSamplerParams()
	if params.Temperature <= 1.0 {
		t.Errorf("Creative Temperature should be > 1.0, got %f", params.Temperature)
	}
	if params.TopK <= 40 {
		t.Errorf("Creative TopK should be > 40, got %d", params.TopK)
	}
}

func TestPreciseSamplerParams(t *testing.T) {
	params := PreciseSamplerParams()
	if params.Temperature >= 0.5 {
		t.Errorf("Precise Temperature should be < 0.5, got %f", params.Temperature)
	}
	if params.TopK >= 40 {
		t.Errorf("Precise TopK should be < 40, got %d", params.TopK)
	}
}

func TestCosineSimilarity(t *testing.T) {
	t.Run("identical vectors", func(t *testing.T) {
		a := []float32{1.0, 0.0, 0.0}
		b := []float32{1.0, 0.0, 0.0}
		sim, err := CosineSimilarity(a, b)
		if err != nil {
			t.Fatalf("Unexpected error: %v", err)
		}
		if math.Abs(float64(sim-1.0)) > 0.001 {
			t.Errorf("Expected similarity ~1.0, got %f", sim)
		}
	})

	t.Run("orthogonal vectors", func(t *testing.T) {
		a := []float32{1.0, 0.0, 0.0}
		b := []float32{0.0, 1.0, 0.0}
		sim, err := CosineSimilarity(a, b)
		if err != nil {
			t.Fatalf("Unexpected error: %v", err)
		}
		if math.Abs(float64(sim)) > 0.001 {
			t.Errorf("Expected similarity ~0.0, got %f", sim)
		}
	})

	t.Run("opposite vectors", func(t *testing.T) {
		a := []float32{1.0, 0.0, 0.0}
		b := []float32{-1.0, 0.0, 0.0}
		sim, err := CosineSimilarity(a, b)
		if err != nil {
			t.Fatalf("Unexpected error: %v", err)
		}
		if math.Abs(float64(sim+1.0)) > 0.001 {
			t.Errorf("Expected similarity ~-1.0, got %f", sim)
		}
	})

	t.Run("mismatched lengths", func(t *testing.T) {
		a := []float32{1.0, 0.0}
		b := []float32{1.0, 0.0, 0.0}
		_, err := CosineSimilarity(a, b)
		if err == nil {
			t.Error("Expected error for mismatched lengths")
		}
	})
}

func TestLoadModelNonexistent(t *testing.T) {
	_, err := LoadModel("/nonexistent/path/model.gguf", nil)
	if err == nil {
		t.Error("Expected error loading nonexistent model")
	}
}

// Tests that require a model file

func TestLoadModel(t *testing.T) {
	if !modelAvailable {
		t.Skip("No test model available")
	}

	model, err := LoadModel(testModelPath, nil)
	if err != nil {
		t.Fatalf("Failed to load model: %v", err)
	}
	defer model.Free()

	if model.NVocab() <= 0 {
		t.Error("NVocab should be positive")
	}
	if model.NEmbd() <= 0 {
		t.Error("NEmbd should be positive")
	}
}

func TestModelProperties(t *testing.T) {
	if !modelAvailable {
		t.Skip("No test model available")
	}

	model, err := LoadModel(testModelPath, nil)
	if err != nil {
		t.Fatalf("Failed to load model: %v", err)
	}
	defer model.Free()

	if model.NCtxTrain() <= 0 {
		t.Error("NCtxTrain should be positive")
	}
	if model.NLayer() <= 0 {
		t.Error("NLayer should be positive")
	}
	if model.NHead() <= 0 {
		t.Error("NHead should be positive")
	}
	if model.Size() <= 0 {
		t.Error("Size should be positive")
	}
	if model.NParams() <= 0 {
		t.Error("NParams should be positive")
	}
	if model.TokenBOS() < 0 {
		t.Error("TokenBOS should be non-negative")
	}
	if model.TokenEOS() < 0 {
		t.Error("TokenEOS should be non-negative")
	}
}

func TestTokenization(t *testing.T) {
	if !modelAvailable {
		t.Skip("No test model available")
	}

	model, err := LoadModel(testModelPath, nil)
	if err != nil {
		t.Fatalf("Failed to load model: %v", err)
	}
	defer model.Free()

	text := "Hello, world!"
	tokens, err := model.Tokenize(text, true, false)
	if err != nil {
		t.Fatalf("Tokenization failed: %v", err)
	}

	if len(tokens) == 0 {
		t.Error("Tokenization should produce tokens")
	}

	decoded, err := model.Detokenize(tokens, false, false)
	if err != nil {
		t.Fatalf("Detokenization failed: %v", err)
	}

	if decoded == "" {
		t.Error("Detokenization should produce text")
	}
}

func TestCreateContext(t *testing.T) {
	if !modelAvailable {
		t.Skip("No test model available")
	}

	model, err := LoadModel(testModelPath, nil)
	if err != nil {
		t.Fatalf("Failed to load model: %v", err)
	}
	defer model.Free()

	ctx, err := NewContext(model, nil)
	if err != nil {
		t.Fatalf("Failed to create context: %v", err)
	}
	defer ctx.Free()

	if ctx.NCtx() <= 0 {
		t.Error("NCtx should be positive")
	}
	if ctx.NBatch() <= 0 {
		t.Error("NBatch should be positive")
	}
}

func TestContextWithParams(t *testing.T) {
	if !modelAvailable {
		t.Skip("No test model available")
	}

	model, err := LoadModel(testModelPath, nil)
	if err != nil {
		t.Fatalf("Failed to load model: %v", err)
	}
	defer model.Free()

	params := ContextParams{
		NCtx:   512,
		NBatch: 256,
	}

	ctx, err := NewContext(model, &params)
	if err != nil {
		t.Fatalf("Failed to create context: %v", err)
	}
	defer ctx.Free()

	if ctx.NCtx() != 512 {
		t.Errorf("Expected NCtx 512, got %d", ctx.NCtx())
	}
	if ctx.NBatch() != 256 {
		t.Errorf("Expected NBatch 256, got %d", ctx.NBatch())
	}
}

func TestGenerateText(t *testing.T) {
	if !modelAvailable {
		t.Skip("No test model available")
	}

	model, err := LoadModel(testModelPath, nil)
	if err != nil {
		t.Fatalf("Failed to load model: %v", err)
	}
	defer model.Free()

	ctx, err := NewContext(model, &ContextParams{NCtx: 256})
	if err != nil {
		t.Fatalf("Failed to create context: %v", err)
	}
	defer ctx.Free()

	params := GreedySamplerParams()
	text, err := ctx.Generate("Hello", 10, &params)
	if err != nil {
		t.Fatalf("Generation failed: %v", err)
	}

	if text == "" {
		t.Error("Generated text should not be empty")
	}
}

func TestClearCache(t *testing.T) {
	if !modelAvailable {
		t.Skip("No test model available")
	}

	model, err := LoadModel(testModelPath, nil)
	if err != nil {
		t.Fatalf("Failed to load model: %v", err)
	}
	defer model.Free()

	ctx, err := NewContext(model, nil)
	if err != nil {
		t.Fatalf("Failed to create context: %v", err)
	}
	defer ctx.Free()

	// Should not panic
	ctx.ClearCache()
}

func TestEmbeddingGenerator(t *testing.T) {
	if !modelAvailable {
		t.Skip("No test model available")
	}

	model, err := LoadModel(testModelPath, nil)
	if err != nil {
		t.Fatalf("Failed to load model: %v", err)
	}
	defer model.Free()

	gen, err := NewEmbeddingGenerator(model, 512, true)
	if err != nil {
		t.Fatalf("Failed to create embedding generator: %v", err)
	}
	defer gen.Free()

	if gen.NEmbd() <= 0 {
		t.Error("NEmbd should be positive")
	}
}
